### ISSUE TITLE
Enable support for DataTypeAttribute

### ISSUE THREAD

**Issue by TimoWillemsen** [Haacked/mvc-metadata-conventions#8]
I've tried making it work with DataTypeAttributes likes [EmailAddress], however it doesn't work.

I think it's because EmailAddress doesn't extend [ValidationAttribute], but DatatypeAttribute instead (which then extends ValidationAttribute).

**Comment by Haacked** [comment]
Closing this unless there's a change that needs to happen in this project.

**Comment by rusanov-vladimir** [comment]
To fix the issue I've change the [line](https://github.com/Haacked/mvc-metadata-conventions/blob/master/ModelMetadataExtensions/ConventionalModelMetadataProvider.cs#L98) to
```
 if (string.IsNullOrEmpty(validationAttribute.ErrorMessage) 
     || validationAttribute.ErrorMessage == ((ValidationAttribute) Activator.CreateInstance(validationAttribute.GetType())).ErrorMessage)
```

It creates some overhead, but at least you don't need to explicitly set message to null.

**Comment by Haacked** [comment]
@rusanov-vladimir thanks. Is this something very specific to the `EmailAddressAttribute`? I wonder if we could special case the fix to that?

**Comment by Haacked** [comment]
Also, I wonder if the issue has been fixed in more recent builds of .NET. Check out this source code comment here: http://referencesource.microsoft.com/#System.ComponentModel.DataAnnotations/DataAnnotations/EmailAddressAttribute.cs,16

My guess is they used to set `ErrorMessage` which causes this problem, but now they set `DefaultErrorMessage`

**Comment by rusanov-vladimir** [comment]
Well, from what I understand the issue happens for all DataTypeAttributes, so yes, it would make sense to special case the fix only for those. 

It happens because in the DataTypeAttribute base class (ValidationAttribute) [ErrorMessage](http://referencesource.microsoft.com/#System.ComponentModel.DataAnnotations/DataAnnotations/ValidationAttribute.cs,adcc113439ab35ba,references) is retrieved this way:

```
 public string ErrorMessage {
            get {
                // DevDiv: 468241
                // If _errorMessage is not set, return the default. This is done to preserve
                // behavior prior to the fix where ErrorMessage showed the non-null message to use.
                return this._errorMessage ?? this._defaultErrorMessage;
            }
...
}
```
ErrorMessage by default is empty, but DefaultErrorMessage [is set for DataAnTypeAttributes](http://referencesource.microsoft.com/#System.ComponentModel.DataAnnotations/DataAnnotations/ValidationAttribute.cs,298e4da9d49b50f0,references).

Now, when previously mentioned line does `string.IsNullOrEmpty(validationAttribute.ErrorMessage)`, for other attributes the check passes, because there is no errorMessage nor defaultErrorMessage set. OTOH for DataTypeAttributes there is DefaultErrorMessage and **ErrorMessage isn't null** or empty.

My fix was to create a new instance of the attribute  and to check if has the same ErrorMessage as the current one. If so, then it's the default ErrorMessage which we can ignore and apply the localized one.

The fix mentioned above `[EmailAddress(ErrorMessage = null)]` works because it's setting defaultErrorMessage to null, so when the check is done ErrorMessage is null and localization is applied.

Hope it makes sense.

**Comment by Haacked** [comment]
@rusanov-vladimir Thanks for the explanation. It makes sense now.

> My fix was to create a new instance of the attribute and to check if has the same ErrorMessage as the current one. If so, then it's the default ErrorMessage which we can ignore and apply the localized one.

Given what you said, I wonder if a cleaner fix is something like:

```csharp
if (string.IsNullOrEmpty(validationAttribute.ErrorMessage) 
     || (validationAttribute as DataTypeAttribute)?.DefaultErrorMessage == validationAttribute.ErrorMessage)
```

Thoughts? And would you like to try and submit this as a pull request so you get credit in the Git history? :smile:

**Comment by rusanov-vladimir** [comment]
Yes, this would be a cleaner fix, if the [DefaultErrorMessage](http://referencesource.microsoft.com/#System.ComponentModel.DataAnnotations/DataAnnotations/ValidationAttribute.cs,298e4da9d49b50f0,references) wouldn't be internal.
So now I would fix it this way:

```
if (string.IsNullOrEmpty(validationAttribute.ErrorMessage) 
     || ( validationAttribute is DataTypeAttribute 
         && validationAttribute.ErrorMessage == 
           ((ValidationAttribute) Activator.CreateInstance(validationAttribute.GetType())).ErrorMessage
        )
)
```
I will provide a PR, if you are OK with this it.
Thank you for a cool solution to localization problem 🥇

**Comment by haacked** [comment]
Fixed by #17. Released in Nuget as [ModelMetadataExtensions v0.1.4](https://www.nuget.org/packages/ModelMetadataExtensions/0.1.4)
